### DISCUSSION TITLE
Multi-channel/device sound

### DISCUSSION THREAD

**Discussion by macdroid53** [FrancescoCeruti/linux-show-player#274] [Q&A] [answered]
Looking at the Qlab screenshot in issue #36 brought to mind a question:

Is is possible to select sound device channels or multiple sound devices.

I'm thinking of a venue that has 5.1 capabilities and wants to send a sound to right rear for some reason.

Or, to a speaker hidden under a set piece so the sound is localized to a particular spot on stage, for example a phone or radio on stage.

**Comment by FrancescoCeruti** [comment]
Unfortunately, AFAIK, GStreamer is a bit limited on this side, unless you write some custom output plugin. The solution is to use JACK, so the answer is yes, if you use the jack-sink and (obviously) a jack server, then, you can manage multi-channel/multi-device, only limitation the LiSP jack-ports are created only when the cue is started (I'm unable to change this behavior).

**Comment by nodiscc** [comment]
I have tested multichannel output with JACK and came to the same conclusion: LiSP's JACK output channels only stay open as long as the cue is playing. When it stops they are removed. Which is a problem since JACK routing is not restored the next time you play the cue.

![cap](https://cloud.githubusercontent.com/assets/3950390/17893920/c37ce1ea-6947-11e6-9848-24992377305a.png)

@FrancescoCeruti is it possible to open a JACK sink as soon as a media cue is created and assigned to jack out, and leave it open?

**Comment by macdroid53** [comment]
That's standard gstreamer behavior, right?

I think there is a way to define which jack port is used as the default connection for gstreamer...but have not figured that out yet. :(

**Comment by FrancescoCeruti** [comment]
the `LiSP out -> Some input-port` can be edited in the jack-sink settings, those are restored, if you edit them in qtjackctl you lose the changes, unfortunately GStreamer doesn't keep the ports alive so I need to use a control client to edit/restore the connections before the stream is played.

**Comment by macdroid53** [comment]
I think there is a system setting that tells gstreamer to look for a particular port, if that is not set, it will connect to the default out (typically, playback 1/2).

If you edit the jack-sink settings in a cue, that would need to be set for every cue a particular sink was desired, correct?

**Comment by FrancescoCeruti** [comment]
By default I left all the port disconnected, this option is provided by GStreamer, then LiSP manage the sink connections using a JACK "control client", the default ports are indeed the playback, or whatever is found under "system" in JACK, and all the available media channels are connected 1-1, 2-2, ..., 8-8.

Yes you need to edit all the cues, this can be done using multi-editing, the jack sink must already be in use in the cues, but you can change the default output plugin (for new cues) in the gstreamer settings.

**Comment by FrancescoCeruti** [comment]
by the way, I've seen your question on linux-audio mailing list, I'm aware of the `port-pattern` property but since I need a client to query the inputs ports for the "pattern" editing I've decided to do the connection myself

**Comment by macdroid53** [comment]
I understand.

But, somewhere over the past few years I figured out how to do this for
another situation.

And it is documented for gstreamer 0.01 on the JACK Audio web site.

But, it seems to have gone missing for gs 1.0, 16.04, and the switch to
dconf.

Mac

On Tue, Aug 23, 2016 at 11:53 AM, Francesco Ceruti <notifications@github.com

**Comment by FrancescoCeruti** [comment]
I've take a look at the [gstjackaudiosink code](https://cgit.freedesktop.org/gstreamer/gst-plugins-good/tree/ext/jack/gstjackaudiosink.c?h=1.6), the auto-connection is done in 2 possible ways:
1. `port-pattern` is not set: query JACK for all the physical and input ports [line 461](https://cgit.freedesktop.org/gstreamer/gst-plugins-good/tree/ext/jack/gstjackaudiosink.c?h=1.6#n461)
2. `port-pattern` is provided: query JACK for the ports that respect the pattern [line 465](https://cgit.freedesktop.org/gstreamer/gst-plugins-good/tree/ext/jack/gstjackaudiosink.c?h=1.6#n464)

Then it go through the `N` stream channels and connect them to a `n` port, if possible. [line 476](https://cgit.freedesktop.org/gstreamer/gst-plugins-good/tree/ext/jack/gstjackaudiosink.c?h=1.6#n476)

There is no evidence of possible way to influence the connections other then `port-pattern`, to solve your problem we need to find a way to configure `port-pattern` system-wide, and even in this case currently you cannot alter the LiSP jackaudiosink connection because the auto-connection is disabled by the application code.

**Comment by macdroid53** [comment, marked as answer]
Playing around with the latest LISP. (please note, this is not a complaint. Being able to connect like this is great!)

My jack is setup as follows:

![jack_setup](https://cloud.githubusercontent.com/assets/6831752/18029387/dd6b605c-6c64-11e6-9773-612c48f9ce0f.png)

When I add a cue in LISP it defaults to JACK, but have to set the connection for each new cue:

![lisp-jack](https://cloud.githubusercontent.com/assets/6831752/18029410/60e9923c-6c65-11e6-969c-7356ba769d16.png)

Is there a way to set the default connection?

Also, if this can't be changed globally, moving from one venue to another could be annoying.

**Reply by clemradio** [reply]
Does that still exist, as of 2023? I’d love to play some cues to a separate output. In LiSP 0.5.1, or the dev version, I don’t see such a feature

**Reply by FrancescoCeruti** [reply]
@clemradio you need to use JACK and configure the cues to use the jack output, [docs here](https://linux-show-player-users.readthedocs.io/en/latest/gst_media_settings.html#output-elements) it should work for both 0.5 and the dev version

**Comment by macdroid53** [comment]
Sorry, hit the wrong button.

**Comment by FrancescoCeruti** [comment]
Currently the only way is to select all the media cues and then multi-edit them.
I'll try to have a global settings option for the jack-sink in the 0.4.1, anyway probably there will be at least one other release for the 0.4.x cycle ;)

**Comment by macdroid53** [comment]
Multi-edit is fine...hmm, well, it would be nice have a select "all media cues", since a show with 100 cues would have action cues mixed in. Selecting (as in ctrl-left click each of the desired cues) one at a time to multi-edit could be error prone and tedious.

**Comment by FrancescoCeruti** [comment]
> Multi-edit is fine...hmm, well, it would be nice have a select "all media cues", since a show with 100 cues would have action cues mixed in. Selecting (as in ctrl-left click each of the desired cues) one at a time to multi-edit could be error prone and tedious.

Done :) take a look at the last commit, you can find it in the `Edit` menu.
